### PR TITLE
Nullable annotation for EFCore/Storage and others

### PR DESCRIPTION
**Pull request by roji**
This is a first PR annotating the EF Core codebase for C# 8.0 nullability. Since this is a huge task, I'm splitting it down into subsections and enabling nullability checks on a file-by-file basis (with `#nullable enable`). When an entire project is annotated, I'll remove all of these and enable checks at the csproj level.

This takes care of:
* EFCore.Abstractions
* EFCore/Storage
* EFCore/Internal
* EFCore/

Part of #14150

Updates the entire solution to use C# 8.0

### CODE REVIEW

File: src/EFCore.Abstractions/DbFunctionAttribute.cs [hunk 1]
```diff
@@ -17,8 +17,8 @@ namespace Microsoft.EntityFrameworkCore
     public class DbFunctionAttribute : Attribute
 #pragma warning restore CA1813 // Avoid unsealed attributes
     {
-        private string _functionName;
-        private string _schema;
+        private string? _functionName;
```
[the remark this answers is not in the document]
Comment (roji) [reply src/EFCore.Abstractions/DbFunctionAttribute.cs]: That's because there's a (public) constructor which doesn't initialize it. I was wondering what that meant - do we infer the store function name somehow from the CLR function name? If it doesn't make any sense, we could consider removing...

[the remark this answers is not in the document]
Comment (smitpatel) [reply src/EFCore.Abstractions/DbFunctionAttribute.cs]: > do we infer the store function name somehow from the CLR function name?

Yes

[the remark this answers is not in the document]
Comment (ajcvickers) [reply src/EFCore.Abstractions/DbFunctionAttribute.cs]: Hmmm. It's actually a bit weird to have something that can be null, but can never be set to null. I would expect this to either be an immutable type, or for the setter to allow null to be set. @smitpatel @divega Any idea why it is this way?

File: src/EFCore/Internal/EntityFinder.cs [hunk 2]
```diff
@@ -50,9 +52,9 @@ public class EntityFinder<TEntity> : IEntityFinder<TEntity>
         ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
         ///     directly from your code. This API may change or be removed in future releases.
         /// </summary>
-        public virtual TEntity Find(object[] keyValues)
+        public virtual TEntity? Find(object[]? keyValues)
         {
-            return keyValues?.Any(v => v == null) != false
+            return keyValues == null || keyValues.Any(v => v == null)
```
Comment (ajcvickers) [inline src/EFCore/Internal/EntityFinder.cs]: Why this change?

File: src/EFCore.Abstractions/DbFunctionAttribute.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (smitpatel) [reply src/EFCore.Abstractions/DbFunctionAttribute.cs]: DbFunctionAttribute.FunctionName can be null (or un-initialized). DbFunction.FunctionName cannot be null.

[the remark this answers is not in the document]
Comment (ajcvickers) [reply src/EFCore.Abstractions/DbFunctionAttribute.cs]: @smitpatel Right, but this is the DbFunctionAttribute class, and it prevents the function name to be set to null, even though null is a valid value here and this class is mutable.

File: src/EFCore/Internal/EntityFinder.cs (hunk 2 above)
Comment (roji) [reply src/EFCore/Internal/EntityFinder.cs]: The compiler currently needs this in order to correctly analyze the flow. It may get better at it in the future, although I personally think the new version is also a bit more readable.

File: src/EFCore/Properties/CoreStrings.Designer.cs [hunk 3]
```diff
@@ -1840,7 +1842,7 @@ public static string NoParameterlessConstructor([CanBeNull] object entityType)
         /// <summary>
         ///     Cannot create a relationship between '{newPrincipalEntityType}.{newPrincipalNavigation}' and '{newDependentEntityType}.{newDependentNavigation}', because there already is a relationship between '{existingPrincipalEntityType}.{existingPrincipalNavigation}' and '{existingDependentEntityType}.{existingDependentNavigation}'. Navigation properties can only participate in a single relationship.
         /// </summary>
-        public static string ConflictingRelationshipNavigation([CanBeNull] object newPrincipalEntityType, [CanBeNull] object newPrincipalNavigation, [CanBeNull] object newDependentEntityType, [CanBeNull] object newDependentNavigation, [CanBeNull] object existingPrincipalEntityType, [CanBeNull] object existingPrincipalNavigation, [CanBeNull] object existingDependentEntityType, [CanBeNull] object existingDependentNavigation)
```
Comment (ajcvickers) [inline src/EFCore/Properties/CoreStrings.Designer.cs]: The designer file for strings here is generated from a T4, so that will likely need updating. /cc @bricelam @AndriySvyryd

File: src/EFCore/Storage/TypeMappingSource.cs [hunk 4]
```diff
@@ -123,7 +125,12 @@ protected TypeMappingSource([NotNull] TypeMappingSourceDependencies dependencies
                         mapping = mapping.Clone(converter);
                     }
 
+#nullable disable
+                    // It seems like if we failed to find a mapping, we insert a null into _explicitMappings.
```
Comment (ajcvickers) [inline src/EFCore/Storage/TypeMappingSource.cs]: Should this be filed as a bug on corefx?

File: src/EFCore/Properties/CoreStrings.Designer.cs (hunk 3 above)
Comment (roji) [reply src/EFCore/Properties/CoreStrings.Designer.cs]: Yeah... I'll take a look at it and will ask for help if needed.

File: src/EFCore/Storage/TypeMappingSource.cs (hunk 4 above)
Comment (roji) [reply src/EFCore/Storage/TypeMappingSource.cs]: This actually turned out to be a mistake on my side - the type for `_explicitMappings` simply wasn't updated allow null values. Pushed a commit fixing this.

File: src/EFCore.Abstractions/DbFunctionAttribute.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (roji) [reply src/EFCore.Abstractions/DbFunctionAttribute.cs]: I'm fine with whatever - the idea was to change/touch nothing beyond the absolutely necessary for nullability, leaving things as they are. Unless you guys really care, we can just merge it as-is, we can always revisit later.

[the remark this answers is not in the document]
Comment (ajcvickers) [reply src/EFCore.Abstractions/DbFunctionAttribute.cs]: No need to do anything in this PR. It's not a big deal.
